ignite/metrics/metrics_lambda.py

```diff
@@ -43,38 +44,65 @@ def __init__(self, f: Callable, *args, **kwargs):
         self.function = f
         self.args = args
         self.kwargs = kwargs
+        self.engine = None
         super(MetricsLambda, self).__init__(device="cpu")
 
     @reinit__is_reduced
     def reset(self) -> None:
-        for i in itertools.chain(self.args, self.kwargs.values()):
-            if isinstance(i, Metric):
-                i.reset()
+        if not self._detach_if_necessary(self.engine):
+            for i in itertools.chain(self.args, self.kwargs.values()):
+                if isinstance(i, Metric):
+                    i.reset()
 
     @reinit__is_reduced
     def update(self, output) -> None:
         # NB: this method does not recursively update dependency metrics,
         # which might cause duplicate update issue. To update this metric,
         # users should manually update its dependencies.
-        pass
+        self._detach_if_necessary(self.engine)
 
     def compute(self) -> Any:
-        materialized = [i.compute() if isinstance(i, Metric) else i for i in self.args]
-        materialized_kwargs = {k: (v.compute() if isinstance(v, Metric) else v) for k, v in self.kwargs.items()}
-        return self.function(*materialized, **materialized_kwargs)
+        if not self._detach_if_necessary(self.engine):
+            materialized = [i.compute() if isinstance(i, Metric) else i for i in self.args]
+            materialized_kwargs = {k: (v.compute() if isinstance(v, Metric) else v) for k, v in self.kwargs.items()}
+            return self.function(*materialized, **materialized_kwargs)
 
     def _internal_attach(self, engine: Engine) -> None:
+        self.engine = weakref.ref(engine)
         for index, metric in enumerate(itertools.chain(self.args, self.kwargs.values())):
             if isinstance(metric, MetricsLambda):
                 metric._internal_attach(engine)
             elif isinstance(metric, Metric):
+                # NB : metrics is attached partially
+                # We must not use is_attached() but rather if these events exist
                 if not engine.has_event_handler(metric.started, Events.EPOCH_STARTED):
                     engine.add_event_handler(Events.EPOCH_STARTED, metric.started)
                 if not engine.has_event_handler(metric.iteration_completed, Events.ITERATION_COMPLETED):
                     engine.add_event_handler(Events.ITERATION_COMPLETED, metric.iteration_completed)
 
     def attach(self, engine: Engine, name: str) -> None:
-        # recursively attach all its dependencies
+        # recursively attach all its dependencies (partially)
         self._internal_attach(engine)
         # attach only handler on EPOCH_COMPLETED
         engine.add_event_handler(Events.EPOCH_COMPLETED, self.completed, name)
+
+    def is_attached(self, engine: Engine) -> bool:
+        self._detach_if_necessary(weakref.ref(engine))
```

Comment:
This line looks a bit strange: `is_attached` is constant method that should not alter the internal state, but `_detach_if_necessary` return a bool that is not used and it also can call `super().detach(engine)`

Reply:
Bug for the missing return, sorry.

My feeling is `is_attached` should not alter _by design_ but not in this concrete implementation. I mean, to know if `MetricsLambda` is attached, it requires check inner metrics and not only `super().is_attached(engine)`. It results from `MetricsLambda`  has an internal state (weakref).
The conflicting scenario tackled by non constant implementation of `is_attached` : 
`super().is_attached() = True` 
but 
`max([m.is_attached(engine) for m in inner_metrics]) = False`.
In that case, `MetricsLambda` is detached during `is_attached` call. I hope it's a little clearer.

Reply:
I think about my previous message. During iterative process of engines, `is_attached` is not used. It could return `super().is_attached() xand max([m.is_attached(engine) for m in inner_metrics])`, so the call could not alter the internal state.